src/context.rs

```diff
@@ -31,14 +36,85 @@ impl Context {
         Context::with_values_and_filters(HashMap::new(), filters)
     }
 
-    pub fn with_values_and_filters(values: HashMap<String, Value>, filters: HashMap<String, Box<Filter>>) -> Context {
+    pub fn with_values_and_filters(values: HashMap<String, Value>,
+                                   filters: HashMap<String, Box<Filter>>) -> Context {
         Context {
-            values: values,
-            filters: filters,
+            stack: vec!(HashMap::new()),
+            globals: values,
+            filters: filters
+        }
+    }
+
+    pub fn add_filter(&mut self, name: &str, filter: Box<Filter>) {
+        self.filters.insert(name.to_owned(), filter);
+    }
+
+    pub fn get_filter<'b>(&'b self, name: &str) -> Option<&'b Box<Filter>> {
+        self.filters.get(name)
+    }
+
+    /// Creates a new variable scope chained to a parent scope.
+    pub fn push_scope(&mut self) {
+        self.stack.push(HashMap::new())
+    }
+
+    /// Removes the topmost stack frame from the local variable stack.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if popping the topmost frame results in an
+    /// empty stack. Given that a context is created with a top-level stack
+    /// frame already in place, empyting the stack should never happen in a
+    /// well-formed program.
+    pub fn pop_scope(&mut self) {
+        if let None = self.stack.pop() {
+            panic!("Pop leaves empty stack")
         }
     }
 
-    /// Gets a value from the rendering context.
+    /// Sets up a new stack frame, executes the supplied function and then
+    /// tears the stack frame down before returning the function's result
+    /// to the caller.
+    ///
+    /// # Examples
+    /// ```
+    /// # use liquid::{Value, Context};
+    /// let mut ctx = Context::new();
+    /// ctx.set_val("test", Value::Num(42f32));
+    /// ctx.with_new_scope(|mut stack_frame| {
+    ///   // stack_frame inherits values from its parent context
+    ///   assert_eq!(stack_frame.get_val("test"), Some(&Value::Num(42f32)));
+    ///
+    ///   // but can (optionally) override them
+    ///   stack_frame.set_local_val("test", Value::Num(3.14f32));
+    ///   assert_eq!(stack_frame.get_val("test"), Some(&Value::Num(3.14f32)));
+    /// });
+    /// // the original value is unchanged once the scope exits
+    /// assert_eq!(ctx.get_val("test"), Some(&Value::Num(42f32)));
+    /// ```
+    pub fn with_new_scope<RvalT, FnT>(&mut self, f: FnT) -> RvalT
```

Comment:
Now, I really like the concept that's behind this function! I think that's an elegant solution.

A small nit: in my terminology of Rust `Context::with_something` would always be used to create a new `Context`. The name confused me for a bit. Maybe we can rename this to e.g. `spawn_scope`? What do you think?